Migrate to symfony/console

weierophinney opened the pull request:
This patch migrates the tooling to use symfony/console instead of zend-stdlib's `ConsoleHelper`. The reasons are:

- We were needing to write argument parsing routines for each and every command, and each differed.
- We were needing to write "help" functionality for each command.
- We were creating a plethora of entry points. We should have one, with many commands available.
- While we could use zf-deploy, using symfony/console will likely get more contributors to the project, due to its ubiquity in the PHP ecosystem. In the end, I also discovered that I ended up writing less code overall in order to create commands, and had better, simpler options for re-use when re-purposing existing command endpoints to re-use the same commands as used by the main entry point.

All existing commands continue to work. I have updated the `expressive` tool to use the following patterns:

- `migrate:*`: the various tools dedicated to migration.
- `module:*`: the various tools dedicated to module manipulation.
- `middleware:*`: the tools dedicated to middleware manipulation
  (currently only creation).

### TODO

- [x] Update unit tests.
- [x] Add deprecation warnings to original, now legacy, scripts.

weierophinney committed:
Migrate to symfony/console

This patch migrates the tooling to use symfony/console instead of
zend-stdlib's `ConsoleHelper`. The reasons are:

- We were needing to write argument parsing routines for each and every
  command, and each differed.
- We were needing to write "help" functionality for each command.
- We were creating a plethora of entry points. We should have one, with
  many commands available.
- While we could use zf-deploy, using symfony/console will likely get
  more contributors to the project, due to its ubiquity in the PHP
  ecosystem. In the end, I also discovered that I ended up writing less
  code overall in order to create commands, and had better, simpler
  options for re-use when re-purposing existing command endpoints to
  re-use the same commands as used by the main entry point.

All existing commands continue to work. I have updated the `expressive`
tool to use the following patterns:

- `migrate:*`: the various tools dedicated to migration.
- `module:*`: the various tools dedicated to module manipulation.
- `middleware:*`: the tools dedicated to middleware manipulation
  (currently only creation).

michalbundyra commented on src/Module/CreateCommand.php:
```diff
@@ -0,0 +1,91 @@
+        } catch (Exception\RuntimeException $e) {
+            $console = $this->getErrorConsole($output);
+            $console->writeln('<error>Error during execution:</error>', true, STDERR);
+            $console->writeln(sprintf('  <error>%s</error>', $ex->getMessage()), true, STDERR);
```
`$ex` is not defined. Here should be `$e` or variable in catch should be renamed (I'd prefer to use `$ex` for exception)

michalbundyra commented on src/Module/CreateCommand.php:
```diff
@@ -0,0 +1,91 @@
+            $output->writeln(sprintf('<info>%s</info>', $message));
+        } catch (Exception\RuntimeException $e) {
+            $console = $this->getErrorConsole($output);
+            $console->writeln('<error>Error during execution:</error>', true, STDERR);
```
Now we have different signature of `writeln` method - from symfony, so there are two params: `$messages` and `$options`.

weierophinney committed:
Initial test restructuring

- Removed obsolete `Command` and `Help` tests
- Re-arranged `Module` tests and updated namespaces and class names

michalbundyra commented on src/Module/DeregisterCommand.php:
```diff
@@ -0,0 +1,67 @@
+            $disable = new Disable($this->projectDir, $modulesPath, $composer);
+            $disable->process($module);
+        } catch (RuntimeException $ex) {
+            $console = $this->getErrorConsole();
```
This method should be called with parameter. It's missing here.

michalbundyra commented on src/Module/RegisterCommand.php:
```diff
@@ -0,0 +1,77 @@
+            $enable->setMoveModuleClass(false);
+            $enable->process($module);
+        } catch (RuntimeException $ex) {
+            $console = $this->getErrorConsole();
```
As above - missing required param.

michalbundyra commented on src/ScanForErrorMiddleware/Scanner.php:
```diff
@@ -34,9 +34,9 @@ class Scanner implements Countable, IteratorAggregate
     /**
      * @param string $path
-     * @param ConsoleHelper $console
+     * @param OutputInterface $console
```
The same should be changed also in line 21.

weierophinney replied:
Fixed.

michalbundyra reviewed (changes requested):
I like update to `symfony/console`! Many things is already done there so it simplify our library :) Couple of things to change and should be fine. Of course updating tests as you pointed is needed. Thanks!

weierophinney committed:
Refactored tests for all commands

This patch updates all tests as follows:

- Removes obsolete `Command` and `Help` tests.
- Adds tests for new symfony/console commands.
- Fixes test setup and expectations for classes that now depend on a
  symfony/console `OutputInterface` instance.
- Fixes issues found during testing, including several also reported by
  @webimpress.

weierophinney committed:
Fixed property typehint

- Per @webimpress

weierophinney commented:
All tests updated and/or new tests written; doing so addressed most of the feedback from @webimpress, and I've updated a property typehint per his feedback as well.

weierophinney commented:
Pinging @GeeH, @akrabat, and @CalEvans for feedback...

Ocramius commented on bin/expressive:
```diff
@@ -10,7 +10,9 @@
-use Zend\Stdlib\ConsoleHelper;
+use Symfony\Component\Console\Application;
+
+const VERSION = '%version%';
```
*cough* PackageVersions *cough*

weierophinney replied:
*cough* PHP 5.6 *cough*

Ocramius commented on bin/expressive:
```diff
@@ -22,6 +24,12 @@ if (file_exists($a = __DIR__ . '/../../../autoload.php')) {
-$return = $command->process(array_slice($argv, 1));
-exit($return);
+$application = new Application('expressive', VERSION);
+$application->add(new CreateMiddleware\CreateMiddlewareCommand('middleware:create'));
```
[`Application#addCommands()`](https://github.com/symfony/symfony/blob/bbb0d5ed424a2215588b65720434a0de34611ea0/src/Symfony/Component/Console/Application.php#L399)

weierophinney replied:
Cool, wasn't aware of that one!

Ocramius commented on bin/expressive-migrate-original-messages:
```diff
@@ -22,6 +24,6 @@ if (file_exists($a = __DIR__ . '/../../../autoload.php')) {
-$command = new Command($argv[0], new ConsoleHelper());
-$return = $command->process(array_slice($argv, 1));
-exit($return);
+$application = new Application('expressive-migrate-original-messages', VERSION);
```
Can include the main CLI app, then set the default command instead.

weierophinney replied:
This is a BC measure; the goal is to eventually remove it.

Also note that the command names differ between the main `expressive` script and this one.

Ocramius commented on bin/expressive-module:
```diff
@@ -22,6 +24,8 @@ if (file_exists($a = __DIR__ . '/../../../autoload.php')) {
-$return = $command->process(array_slice($argv, 1));
-exit($return);
+$application = new Application('expressive-module', VERSION);
+$application->add(new CreateCommand('create'));
```
Was this done just to keep the command names BC compliance?

weierophinney replied:
Yes. Since we already have documented migration steps, I cannot break how these original scripts worked; hence the difference between command names and what commands are exposed.

Ocramius commented on bin/expressive-pipeline-from-config:
```diff
@@ -22,6 +24,6 @@ if (file_exists($a = __DIR__ . '/../../../autoload.php')) {
-$command = new Command($argv[0], new ConsoleHelper());
-$return = $command->process(array_slice($argv, 1));
-exit($return);
+$application = new Application('expressive-pipeline-from-config', VERSION);
```
Same as above: main app, set default command in this file

weierophinney replied:
See above; this is a BC measure.

Ocramius commented on bin/expressive-scan-for-error-middleware:
```diff
@@ -22,6 +24,6 @@ if (file_exists($a = __DIR__ . '/../../../autoload.php')) {
-$command = new Command($argv[0], new ConsoleHelper());
-$return = $command->process(array_slice($argv, 1));
-exit($return);
+$application = new Application('expressive-scan-for-error-middleware', VERSION);
```
Same as above: main app, set default command in this file

weierophinney replied:
See above; this is a BC measure.

Ocramius commented on composer.json:
```diff
@@ -43,7 +47,6 @@
     },
     "bin": [
         "bin/expressive",
-        "bin/expressive-create-middleware",
```
Possible BC break? Or do we not want to go there?

weierophinney replied:
Never before released; that feature was for the upcoming 0.4.0 release. As such, I'd rather _not_ release that script, and instead ship it only as a command in the `expressive` script.

Ocramius commented on composer.json:
```diff
@@ -9,13 +9,17 @@
     },
     "require": {
         "php": "^5.6 || ^7.0",
+        "symfony/console": "^3.2",
```
Please also specify the current patch version used: we don't want people to start with 3.2.0, but rather 3.2.7 (current)

Ocramius commented on src/ErrorConsoleTrait.php:
```diff
@@ -0,0 +1,27 @@
+     * @param OutputInterface $output
+     * @return OutputInterface
+     */
+    private function getErrorConsole(OutputInterface $output)
```
All this fuss just for some errors? Nah, just let the exceptions bubble up, and consumers can use `-vvv` to see all the details

Ocramius replied:
If really needed, can be made into a `public static function` instead (better, and testable)

Ocramius commented on src/GenerateProgrammaticPipelineFromConfig/Generator.php:
```diff
@@ -240,10 +239,11 @@ private function generatePipeline(array $config)
 
             if ($error) {
-                $this->console->writeLine(sprintf(
+                $console = $this->getErrorConsole($this->console);
```
Would suggest just printing to `$this->console`, unless STDOUT is important in this generator

Ocramius commented on src/CreateMiddleware/CreateMiddlewareCommand.php:
```diff
@@ -0,0 +1,92 @@
+            $console = $this->getErrorConsole($output);
+            $console->writeln('<error>Error during generation:</error>');
+            $console->writeln(sprintf('  <error>%s</error>', $e->getMessage()));
+            return 1;
```
Let the exception bubble up, otherwise you are making it really hard for consumers to see why the `$generator` may have failed, as no stack trace is produced

Ocramius commented on src/GenerateProgrammaticPipelineFromConfig/PipelineFromConfigCommand.php:
```diff
@@ -0,0 +1,115 @@
+            $console = $this->getErrorConsole($output);
+            $console->writeln('<error>Error during generation:</error>');
+            $console->writeln(sprintf('  <error>%s</error>', $e->getMessage()));
+            return 1;
```
Same: let the exception bubble up instead

Ocramius commented on src/MigrateOriginalMessageCalls/MigrateOriginalMessageCallsCommand.php:
```diff
@@ -0,0 +1,129 @@
+                '<error>Unable to determine src directory: %s</error>',
+                $e->getMessage()
+            ));
+            return 1;
```
Let exceptions bubble up. If the exception message is meaningful, no additional code is needed

Ocramius commented on src/Module/CreateCommand.php:
```diff
@@ -0,0 +1,91 @@
+            $creation = new Create();
+            $message = $creation->process($module, $modulesPath, $this->projectDir);
+            $output->writeln(sprintf('<info>%s</info>', $message));
+        } catch (Exception\RuntimeException $e) {
```
remove, let it bubble up

Ocramius commented on src/Module/DeregisterCommand.php:
```diff
@@ -0,0 +1,72 @@
+        try {
+            $disable = new Disable($this->projectDir, $modulesPath, $composer);
+            $disable->process($module);
+        } catch (RuntimeException $ex) {
```
Remove, let it bubble up

Ocramius commented on src/Module/RegisterCommand.php:
```diff
@@ -0,0 +1,77 @@
+            $enable = new Enable($this->projectDir, $modulesPath, $composer);
+            $enable->setMoveModuleClass(false);
+            $enable->process($module);
+        } catch (RuntimeException $ex) {
```
same

Ocramius commented on src/ScanForErrorMiddleware/ScanForErrorMiddlewareCommand.php:
```diff
@@ -0,0 +1,130 @@
+    {
+        try {
+            $src = $this->getSrcDir($input);
+        } catch (ArgvException $e) {
```
same

weierophinney committed:
Use `addCommands` instead of multiple `addCommand` calls

- per @Ocramius

weierophinney committed:
Use latest patch release of symfony/console

- per @Ocramius

weierophinney committed:
Write all migrate:pipeline output to STDOUT

- Per @Ocramius

GeeH commented:
The move to this library is amazing, we should always look to use community defacto libraries instead of having to code and support our own whatever the vendor namespace. I'm happy.

I don't have time to review the code sorry (and I don't know the library very well anyway), but...

![image](https://cloud.githubusercontent.com/assets/613376/24898328/732f72d6-1e94-11e7-894a-f8c6fc8cc666.png)

weierophinney committed:
Do not worry about error output

Per @Ocramius, it's simpler to allow exceptions to bubble up, as
symfony/console will report the exception message always, and, when
verbosity is enabled, more details including the stack trace.

weierophinney commented:
@Ocramius I've incorporated your feedback; thanks a ton!

weierophinney committed:
Updated CHANGELOG for #23

Combines the entry for #22 with #23.

weierophinney committed:
Added tooling for bumping the release version of the scripts.

weierophinney committed:
Added deprecation notices to each of the original scripts

weierophinney committed:
Better names for test cases that expect exceptions

weierophinney committed:
Simplify `bump-version dev`

Have it invoke `to` with a fixed `version` argument.

Ocramius commented on src/GenerateProgrammaticPipelineFromConfig/PipelineFromConfigCommand.php:
```diff
@@ -64,16 +61,9 @@ protected function execute(InputInterface $input, OutputInterface $output)
-            return 1;
-        }
+        $generator = new Generator($output);
+        $generator->projectDir = $this->projectDir;
```
Argh... Missed this before, why is this necessary?

weierophinney replied:
Actually... no longer necessary. It was necessary before I started mocking the various internal classes, but now it isn't. I'll nuke it in the various `*Command` classes.

It's still necessary for the internal classes that do filesystem operations, though I _may_ be able to move that to constructor arguments. I'll review.

Ocramius commented on src/Module/CommandCommonTrait.php:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * Add default arguments and options used by all commands.
+     */
+    private function addDefaultOptionsAndArguments()
```
This should be a `public static function`, referenced by classes using it.

Ocramius commented on src/Module/CommandCommonTrait.php:
```diff
@@ -0,0 +1,59 @@
+     * @param InputInterface $input
+     * @return string
+     */
+    private function getModulesPath(InputInterface $input)
```
This should be a `public static function`, referenced by classes using it.

Ocramius commented on src/Module/CommandCommonTrait.php:
```diff
@@ -0,0 +1,59 @@
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+
+trait CommandCommonTrait
```
To be moved to a static class and marked as `@internal`, since we really don't want to distribute this as public/reusable API

weierophinney replied:
Fair enough; I'll make that change (and the ones you suggest below) momentarily.

weierophinney committed:
Remove public `$projectDir` wherever possible

This includes all commands, though a few, where we needed to test path
existence, required the ability to inject a path, which is now handled
by a setter; when the setter is not called, `getcwd()` is then assumed.

For internal classes, this is now handled as either a passed argument,
or via a constructor argument.

weierophinney committed:
Refactor CommandCommonTrait to final, static class CommandCommonOptions

- Per @Ocramius

Ocramius commented:
:ship:
